File: pkg/protocol/protocol.go
package protocol

import (
	"encoding/json"
	"fmt"

	"excelsior/pkg/llm"
)

// Ver is the protocol version string applied to every [Envelope].
const Ver = "v1"

// Envelope is the WS frame.
type Envelope struct {
	Ver     string          `json:"ver"`
	ID      string          `json:"id,omitempty"`
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Decode unmarshals Payload into v. Returns nil if payload is empty.
func (e Envelope) Decode(v any) error {
	if len(e.Payload) == 0 {
		return nil
	}
	if err := json.Unmarshal(e.Payload, v); err != nil {
		return &ProtocolError{
			Op:      "decode",
			MsgType: e.Type,
			Ver:     e.Ver,
			Err:     fmt.Errorf("%w: %v", ErrInvalidPayload, err),
		}
	}
	return nil
}

// MarshalPayload marshals v to json.RawMessage safely without panicking.
// Returns nil, nil if v is nil.
func MarshalPayload(v any) (json.RawMessage, error) {
	if v == nil {
		return nil, nil
	}
	b, err := json.Marshal(v)
	if err != nil {
		return nil, &ProtocolError{
			Op:  "marshal",
			Err: fmt.Errorf("%w: %v", ErrInvalidPayload, err),
		}
	}
	return b, nil
}

// NewEnvelope creates a versioned envelope with JSON payload.
func NewEnvelope(typ string, payload any) Envelope {
	if payload == nil {
		return Envelope{Ver: Ver, Type: typ}
	}
	b, err := json.Marshal(payload)
	if err != nil {
		return Envelope{Ver: Ver, Type: typ}
	}
	return Envelope{Ver: Ver, Type: typ, Payload: b}
}

// NewEnvelopeWithID creates a versioned envelope with ID and payload.
func NewEnvelopeWithID(id, typ string, payload any) Envelope {
	if payload == nil {
		return Envelope{Ver: Ver, ID: id, Type: typ}
	}
	b, err := json.Marshal(payload)
	if err != nil {
		return Envelope{Ver: Ver, ID: id, Type: typ}
	}
	return Envelope{Ver: Ver, ID: id, Type: typ, Payload: b}
}

// Message types for [Envelope.Type].
const (
	TypeChatReq            = "chat.req"
	TypeDelta              = "delta"
	TypeDone               = "done"
	TypeError              = "error"
	TypeAskReq             = "ask.req"
	TypeAskResp            = "ask.resp"
	TypePermissionReq      = "permission.req"
	TypePermissionResp     = "permission.resp"
	TypePing               = "ping"
	TypePong               = "pong"
	TypeSessionList        = "session.list"
	TypeSessionData        = "session.data"
	TypeSessionCreate      = "session.create"
	TypeSessionDelete      = "session.delete"
	TypeSessionRename      = "session.rename"
	TypeSessionSubscribe   = "session.subscribe"
	TypeSessionUnsubscribe = "session.unsubscribe"
	TypeWorkspaceSet       = "workspace.set"
	TypeSettingsGet        = "settings.get"
	TypeSettingsSet        = "settings.set"
)

// WorkspaceSetReq sets active workspace path in engine
type WorkspaceSetReq struct {
	Workspace string `json:"workspace"`
}

// ChatReq is client → engine to start a turn. SessionID ties it to a session sidebar entry.
type ChatReq struct {
	SessionID string        `json:"sessionId,omitempty"`
	Model     string        `json:"model"`
	Messages  []llm.Message `json:"messages"`
}

// Delta is engine → client streaming fragment (maps to agent.StreamEvent).
type Delta struct {
	Type         string `json:"type"` // text|reasoning|tool_start|tool_result|done|error
	Text         string `json:"text,omitempty"`
	Reasoning    string `json:"reasoning,omitempty"`
	ToolName     string `json:"toolName,omitempty"`
	ToolCallID   string `json:"toolCallID,omitempty"`
	ToolArgs     string `json:"toolArgs,omitempty"`
	ToolResult   string `json:"toolResult,omitempty"`
	FinishReason string `json:"finishReason,omitempty"`
	PromptTokens int    `json:"promptTokens,omitempty"`
	CompletionTokens int `json:"completionTokens,omitempty"`
	TotalTokens  int    `json:"totalTokens,omitempty"`
}

// AskReq is engine → client when agent calls askQuestion.
type AskReq struct {
	Question string   `json:"question"`
	Options  []string `json:"options"` // 3
}

// AskResp is client → engine with user choice.
type AskResp struct {
	Selected int    `json:"selected"` // 0..2 or -1 for manual
	Answer   string `json:"answer"`
	Label    string `json:"label"`
}

// PermissionReq is engine → client when agent requests mutating operation approval.
type PermissionReq struct {
	Tool     string `json:"tool"` // "write" | "edit" | "bash"
	FilePath string `json:"filePath,omitempty"`
	Preview  string `json:"preview,omitempty"`
	Command  string `json:"command,omitempty"`
}

// PermissionResp is client → engine with user decision.
type PermissionResp struct {
	Approved bool `json:"approved"`
}

// Session management (unified for TUI/web/desktop).

// SessionListReq requests the session list.
type SessionListReq struct{}

// SessionListResp contains the session list.
type SessionListResp struct {
	Sessions []SessionInfo `json:"sessions"`
}

// SessionInfo is a summary entry for the session sidebar.
type SessionInfo struct {
	ID        string `json:"id"`
	Title     string `json:"title"` // first user message truncated
	UpdatedAt string `json:"updatedAt"`
	Count     int    `json:"count"`
}

// SessionCreateReq creates a new session.
type SessionCreateReq struct {
	Title string `json:"title,omitempty"`
}

// SessionCreateResp is the result of [TypeSessionCreate].
type SessionCreateResp struct {
	ID string `json:"id"`
}

// SessionDeleteReq deletes a session.
type SessionDeleteReq struct {
	ID string `json:"id"`
}

// SessionDataReq requests the full message history for a session.
type SessionDataReq struct {
	ID string `json:"id"`
}

// SessionDataResp contains the full message history for a session.
type SessionDataResp struct {
	ID       string        `json:"id"`
	Messages []llm.Message `json:"messages"`
}

// SessionRenameReq renames a session.
type SessionRenameReq struct {
	ID    string `json:"id"`
	Title string `json:"title"`
}

// SessionSubscriptionReq changes the sessions whose live events a connection receives.
type SessionSubscriptionReq struct {
	ID string `json:"id"`
}

// SettingsGetReq requests current settings (e.g. permission mode).
type SettingsGetReq struct{}

// SettingsGetResp returns current settings.
type SettingsGetResp struct {
	Permission string `json:"permission"` // ask|allow|deny
	AllowAll   bool   `json:"allowAll"`
}

// SettingsSetReq updates settings. Only provided fields are updated.
type SettingsSetReq struct {
	Permission *string `json:"permission,omitempty"`
	AllowAll   *bool   `json:"allowAll,omitempty"`
}

// SettingsSetResp confirms updated settings.
type SettingsSetResp struct {
	Permission string `json:"permission"`
	AllowAll   bool   `json:"allowAll"`
}

File: apps/electron/lib/protocol.ts
// Wire contract: pkg/protocol. Go fixtures exercise this decoder in npm test.
export const RUN_CAPABILITY = "run-lifecycle-v1";
export type Envelope<T = unknown> = { ver: "v1"; workspace?: string; id?: string; type: string; payload: T };
export type Message = { role: string; content?: string; reasoning_content?: string; tool_call_id?: string; name?: string; tool_calls?: { id: string; type: string; function: { name: string; arguments: string } }[] };
export type ChatReq = { sessionId?: string; model: string; messages: Message[] };
export type Delta = { runId: string; sessionId: string; type: string; text?: string; reasoning?: string; toolName?: string; toolCallID?: string; toolArgs?: string; toolResult?: string; finishReason?: string; promptTokens?: number; completionTokens?: number; totalTokens?: number };
export type SessionUsage = { prompt: number; completion: number; total: number };
export type AskReq = { runId: string; interactionId: string; sessionId: string; question: string; options: string[] };
export type AskResp = Pick<AskReq, "runId" | "interactionId" | "sessionId"> & { selected: number; answer: string; label: string };
export type PermissionReq = { runId: string; interactionId: string; sessionId: string; tool: string; filePath?: string; preview?: string; command?: string };
export type PermissionResp = Pick<PermissionReq, "runId" | "interactionId" | "sessionId"> & { approved: boolean };
export type SessionInfo = { id: string; title: string; count: number; updatedAt?: string; branch?: string; added?: number; deleted?: number };
export type SettingsGetResp = { permission: string; allowAll: boolean };
export type SettingsSetResp = SettingsGetResp;
export type Outcome = { sessionId: string; runId: string; status: "succeeded" | "failed" | "canceled" | "persistence_failed"; persisted: boolean; error?: string; code?: string };
export type Snapshot = { id: string; runId?: string; running: boolean; messages: Message[]; events?: Delta[]; pending?: Envelope<AskReq | PermissionReq>; outcome?: Outcome; unsavedAvailable: boolean; projectionUnavailable: boolean };
export type Responses = {
 "auth": { ok: boolean; workspace: string; capabilities: string[] };
 "workspace.set": { workspace: string };
 "workspace.status": { branch: string };
 "session.list": { sessions: SessionInfo[] };
 "session.data": Snapshot;
 "session.create": { id: string };
 "session.delete": { deleted: string };
 "session.rename": { id: string; title: string };
 "session.unsubscribe": { id: string };
 "chat.cancel": { sessionId: string; runId: string; accepted: boolean };
 "ask.resp": { accepted: boolean };
 "permission.resp": { accepted: boolean };
 "settings.get": SettingsGetResp;
 "settings.set": SettingsSetResp;
 "delta": Delta;
 "done": Outcome;
 "ask.req": AskReq;
 "permission.req": PermissionReq;
 "interaction.done": { sessionId: string; runId: string; interactionId: string };
 "error": { error: string; code: string; sessionId?: string; runId?: string };
 "pong": null;
};
export type WireMessage = { [K in keyof Responses]: Envelope<Responses[K]> & { type: K } }[keyof Responses];
export type Commands = {
 "workspace.set": { workspace: string };
 "workspace.status": Record<string, never>;
 "session.list": Record<string, never>;
 "settings.get": Record<string, never>;
 "settings.set": { permission?: string; allowAll?: boolean };
 "session.create": { title?: string };
 "session.data": { id: string };
 "session.delete": { id: string };
 "session.rename": { id: string; title: string };
 "session.subscribe": { id: string };
 "session.unsubscribe": { id: string };
 "chat.req": ChatReq;
 "chat.cancel": { sessionId: string; runId: string };
 "ask.resp": AskResp;
 "permission.resp": PermissionResp;
};
export type ReplyType<K extends keyof Commands> = K extends "chat.req" | "session.subscribe" ? "session.data" : K extends keyof Responses ? K : never;

function object(v: unknown): Record<string, unknown> {
 if (!v || typeof v !== "object" || Array.isArray(v)) throw new Error("Invalid engine object");
 return v as Record<string, unknown>;
}
function strings(p: Record<string, unknown>, keys: string[], optional = false) {
 for (const key of keys) if (!(optional && p[key] === undefined) && typeof p[key] !== "string") throw new Error("Invalid engine field: " + key);
}
function bools(p: Record<string, unknown>, keys: string[]) {
 for (const key of keys) if (typeof p[key] !== "boolean") throw new Error("Invalid engine flag: " + key);
}
function array(v: unknown): unknown[] { if (!Array.isArray(v)) throw new Error("Invalid engine array"); return v; }
function identity(p: Record<string, unknown>) { strings(p, ["sessionId", "runId"]); }
function outcome(v: unknown) {
 const p = object(v); identity(p); bools(p, ["persisted"]); strings(p, ["error", "code"], true);
 if (!["succeeded", "failed", "canceled", "persistence_failed"].includes(String(p.status))) throw new Error("Invalid run outcome");
 if (p.status === "succeeded" && p.persisted !== true) throw new Error("Persistent engine success was not saved");
}
function delta(v: unknown) {
 const p = object(v); identity(p); strings(p, ["type"]);
 strings(p, ["text", "reasoning", "toolName", "toolCallID", "toolArgs", "toolResult", "finishReason"], true);
 for (const k of ["promptTokens", "completionTokens", "totalTokens"]) if (p[k] !== undefined && (typeof p[k] !== "number" || !Number.isFinite(p[k]) || p[k] < 0)) throw new Error("Invalid token usage");
}
function interaction(type: string, v: unknown) {
 const p = object(v); identity(p); strings(p, ["interactionId"]);
 if (type === "ask.req") { strings(p, ["question"]); if (p.options === null) p.options = []; for (const x of array(p.options)) if (typeof x !== "string") throw new Error("Invalid option"); }
 else { strings(p, ["tool"]); strings(p, ["filePath", "preview", "command"], true); }
}
function snapshot(v: unknown) {
 const p = object(v); strings(p, ["id"]); strings(p, ["runId"], true); bools(p, ["running", "unsavedAvailable", "projectionUnavailable"]);
 if (p.running && !p.runId) throw new Error("Running snapshot lacks run identity");
 for (const v of array(p.messages)) {
  const m = object(v); strings(m, ["role"]); strings(m, ["content", "reasoning_content", "tool_call_id", "name"], true);
  if (!["user", "assistant", "system", "tool"].includes(String(m.role))) throw new Error("Invalid message role");
  for (const v of m.tool_calls === undefined ? [] : array(m.tool_calls)) { const t = object(v); strings(t, ["id", "type"]); strings(object(t.function), ["name", "arguments"]); }
 }
 for (const e of p.events === undefined ? [] : array(p.events)) { delta(e); const d = object(e); if (d.sessionId !== p.id || d.runId !== p.runId) throw new Error("Snapshot event scope mismatch"); }
 if (p.outcome) { outcome(p.outcome); const o = object(p.outcome); if (o.sessionId !== p.id || o.runId !== p.runId) throw new Error("Outcome scope mismatch"); }
 if (p.pending) {
  const e = object(p.pending);
  if (e.type !== "ask.req" && e.type !== "permission.req") throw new Error("Invalid pending interaction");
  interaction(e.type, e.payload); const i = object(e.payload);
  if (i.sessionId !== p.id || i.runId !== p.runId || !p.running) throw new Error("Interaction scope mismatch");
 }
}
export function decodeMessage(raw: string): WireMessage {
 if (raw.length > 32 * 1024 * 1024) throw new Error("Engine frame too large");
 const e = object(JSON.parse(raw)); strings(e, ["type"]); strings(e, ["id", "workspace"], true);
 if (e.ver !== "v1") throw new Error("Unsupported engine protocol");
 if (e.type === "pong") return e as WireMessage;
 const p = object(e.payload);
 if (e.type !== "auth" && typeof e.workspace !== "string") throw new Error("Missing workspace scope");
 switch (e.type) {
 case "auth": bools(p, ["ok"]); strings(p, ["workspace"]); for (const c of array(p.capabilities)) if (typeof c !== "string") throw new Error("Invalid capability"); break;
 case "workspace.set": strings(p, ["workspace"]); break;
 case "workspace.status": strings(p, ["branch"]); break;
 case "session.list":
  for (const v of array(p.sessions)) { const s = object(v); strings(s, ["id", "title"]); strings(s, ["updatedAt", "branch"], true); if (typeof s.count !== "number") throw new Error("Invalid count"); } break;
 case "session.data": snapshot(p); break;
 case "session.create": case "session.unsubscribe": strings(p, ["id"]); break;
 case "session.delete": strings(p, ["deleted"]); break;
 case "session.rename": strings(p, ["id", "title"]); break;
 case "settings.get": case "settings.set": strings(p, ["permission"]); bools(p, ["allowAll"]); break;
 case "chat.cancel": identity(p); bools(p, ["accepted"]); break;
 case "ask.resp": case "permission.resp": bools(p, ["accepted"]); break;
 case "delta": delta(p); break;
 case "done": outcome(p); break;
 case "ask.req": case "permission.req": interaction(e.type, p); break;
 case "interaction.done": identity(p); strings(p, ["interactionId"]); break;
 case "error": strings(p, ["error", "code"]); strings(p, ["sessionId", "runId"], true); break;
 default: throw new Error("Unknown engine message: " + e.type);
 }
 return e as WireMessage;
}
